src/box/iproto.cc

```diff
@@ -3312,6 +3450,20 @@ iproto_set_msg_max(int new_iproto_msg_max)
 	}
 }
 
+/**
+ * Notifies IPROTO threads that a new request handler has been set.
+ */
+static void
+iproto_cfg_override(uint32_t req_type, bool is_set)
+{
+	struct iproto_cfg_msg cfg_msg;
+	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_OVERRIDE);
+	cfg_msg.override.req_type = req_type;
+	cfg_msg.override.is_set = is_set;
+	for (int i = 0; i < iproto_threads_count; ++i)
+		iproto_do_cfg_crit(&iproto_threads[i], &cfg_msg);
```

Comment:
> The worst case will be when TX thread updated half IProto threads with a new override, then TX hangs (due to long select, for example). Then some requests of this type will be overridden, some will not. Simply depending on which connection you will use.

TBH, I don't really understand this particular comment: AFAIU, IPROTO configuration is done synchronously, i.e., TX thread waits until the configuration message is delivered to all IPROTO threads.

Reply:
There is no such thing as "TX thread waits". We can't afford blocking the whole single main thread for anything. Instead, it consists of fibers, which can be waiting. But while one fiber waits, others keep working. I was able to make a test which you can find below. It is not fair though - it assumes that one `cbus_call()` (= `iproto_do_cfg_crit()`) might yield multiple times before/after it sends anything to another thread. I couldn't find an easy way to do the same with `usleep()`. Hence you can do whatever you want with this bug, I won't block the patch or anything.

```Diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index 561c00fa1..b8045f9d7 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -3490,8 +3490,14 @@ iproto_cfg_override(uint32_t req_type, bool is_set)
 	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_OVERRIDE);
 	cfg_msg.override.req_type = req_type;
 	cfg_msg.override.is_set = is_set;
-	for (int i = 0; i < iproto_threads_count; ++i)
+	for (int i = 0; i < iproto_threads_count; ++i) {
+		ERROR_INJECT_COUNTDOWN(ERRINJ_IPROTO_CFG_DELAY_COUNTDOWN, {
+			struct errinj *e = errinj(ERRINJ_IPROTO_CFG_DELAY, ERRINJ_BOOL);
+			e->bparam = true;
+			ERROR_INJECT_YIELD(ERRINJ_IPROTO_CFG_DELAY);
+		});
 		iproto_do_cfg_crit(&iproto_threads[i], &cfg_msg);
+	}
 }
 
 int
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index 6aa475f12..220dc565f 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -181,6 +181,8 @@ struct errinj {
 	_(ERRINJ_XLOG_READ, ERRINJ_INT, {.iparam = -1}) \
 	_(ERRINJ_XLOG_RENAME_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_MAIN_MAKE_FILE_ON_RETURN, ERRINJ_BOOL, {.bparam = false}) \
+	_(ERRINJ_IPROTO_CFG_DELAY_COUNTDOWN, ERRINJ_INT, {.iparam = -1}) \
+	_(ERRINJ_IPROTO_CFG_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 
 ENUM0(errinj_id, ERRINJ_LIST);
```
```Lua
net = require('net.box')
log = require('log')
fiber = require('fiber')
box.cfg{
    listen = 3313,
    iproto_threads = 2,
}
box.schema.user.grant('guest', 'super')

conns = {}
for i = 1, 100 do
    table.insert(conns, net.connect(3313, {wait_connected = false}))
end
for _, c in pairs(conns) do
    c:wait_connected()
end

is_intercepted = nil
prev_intercept = nil
conflict_count = 0

f = fiber.create(function()
    box.error.injection.set("ERRINJ_IPROTO_CFG_DELAY_COUNTDOWN", 1)
    box.iproto.override(box.iproto.type.PING, function()
        log.info('intercept')
        is_intercepted = true
        return false
    end)
end)
fiber.sleep(1)
results = {}
for _, c in pairs(conns) do
    is_intercepted = false
    c:ping()
    if prev_intercept == nil then
        prev_intercept = is_intercepted
    elseif prev_intercept ~= is_intercepted then
        conflict_count = conflict_count + 1
    end
    prev_intercept = is_intercepted
    table.insert(results, is_intercepted)
end
```
Then print those 2 variables:
```Lua
results
conflict_count
```
The idea is to make 2 iproto threads, each gets some connections. Then I start an override. There is a moment when one thread is overridden and another is not. I can make synchronous sequential requests and get some of them intercepted, some of them not. Although to be fair, inside single connection this won't happen. This is only observable across multiple connections (because they are pinned to threads).

`conflict_count` shows how many times the interception was on and off inconsistently. In `results` you can see that the results are like `true, true, false, false, true, true, ...` etc.

Reply:
Discussed with @locker: the main use case will be setting up the override set before server configuration, so the problem you describe will not occur there.

Reply:
> There is no such thing as "TX thread waits". We can't afford blocking the whole single main thread for anything. Instead, it consists of fibers, which can be waiting. But while one fiber waits, others keep working.

Thanks for pointing this out!

Reply:
> There will be a problem with this approach (copy the override table in each thread): [#5990 (comment)](https://github.com/tarantool/tarantool/issues/5990#issuecomment-1367943368). I won't block the patch because of it though. Just noting it here.

Let's ignore this problem for now, because it looks minor and it'd be too difficult to fix. We'll open a ticket for it if anybody complains.